// @asymmetrik/ngx-leaflet/__ivy_ngcc__/dist/leaflet/leaflet.module.js
import { ModuleWithProviders, NgModule } from '@angular/core';
import { LeafletDirective } from './core/leaflet.directive';
import { LeafletLayerDirective } from './layers/leaflet-layer.directive';
import { LeafletLayersDirective } from './layers/leaflet-layers.directive';
import { LeafletLayersControlDirective } from './layers/control/leaflet-control-layers.directive';
import { LeafletBaseLayersDirective } from './layers/base/leaflet-baselayers.directive';
import * as ɵngcc0 from '@angular/core';
var LeafletModule = /** @class */ (function () {
    function LeafletModule() {
    }
    LeafletModule.forRoot = function () {
        return { ngModule: LeafletModule, providers: [] };
    };
    /** @nocollapse */
    LeafletModule.ctorParameters = function () { return []; };
LeafletModule.ɵfac = function LeafletModule_Factory(t) { return new (t || LeafletModule)(); };
LeafletModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: LeafletModule });
LeafletModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LeafletModule, [{
        type: NgModule,
        args: [{
                exports: [
                    LeafletDirective,
                    LeafletLayerDirective,
                    LeafletLayersDirective,
                    LeafletLayersControlDirective,
                    LeafletBaseLayersDirective
                ],
                declarations: [
                    LeafletDirective,
                    LeafletLayerDirective,
                    LeafletLayersDirective,
                    LeafletLayersControlDirective,
                    LeafletBaseLayersDirective
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LeafletModule, { declarations: function () { return [LeafletDirective,
        LeafletLayerDirective,
        LeafletLayersDirective,
        LeafletLayersControlDirective,
        LeafletBaseLayersDirective]; }, exports: function () { return [LeafletDirective,
        LeafletLayerDirective,
        LeafletLayersDirective,
        LeafletLayersControlDirective,
        LeafletBaseLayersDirective]; } }); })();
    return LeafletModule;
}());
export { LeafletModule };
